Build failure: python3Packages.llm

### Steps To Reproduce

Steps to reproduce the behavior:
1. `nix build github:NixOS/nixpkgs/b7546315457a024bdd2471417bdf8690423c58fb#python3Packages.llm`

### Build log

```
Sourcing python-remove-tests-dir-hook
Sourcing python-catch-conflicts-hook.sh
Sourcing python-remove-bin-bytecode-hook.sh
Sourcing pypa-build-hook
Using pypaBuildPhase
Sourcing python-runtime-deps-check-hook
Using pythonRuntimeDepsCheckHook
Sourcing pypa-install-hook
Using pypaInstallPhase
Sourcing python-imports-check-hook.sh
Using pythonImportsCheckPhase
Sourcing python-namespaces-hook
Sourcing python-catch-conflicts-hook.sh
Sourcing pytest-check-hook
Using pytestCheckPhase
Running phase: unpackPhase
@nix { "action": "setPhase", "phase": "unpackPhase" }
unpacking source archive /nix/store/gjl2sqyryi980ilajilcbrcm6ybi5987-source
source root is source
setting SOURCE_DATE_EPOCH to timestamp 315619200 of file source/tests/test_templates.py
Running phase: patchPhase
@nix { "action": "setPhase", "phase": "patchPhase" }
applying patch /nix/store/xpvy4c3qmhn6p23bbhngvhm9ldg8aajx-001-disable-install-uninstall-commands.patch
patching file llm/cli.py
Hunk #1 succeeded at 1034 (offset 20 lines).
Hunk #2 succeeded at 1042 (offset 20 lines).
Running phase: updateAutotoolsGnuConfigScriptsPhase
@nix { "action": "setPhase", "phase": "updateAutotoolsGnuConfigScriptsPhase" }
Running phase: configurePhase
@nix { "action": "setPhase", "phase": "configurePhase" }
no configure script, doing nothing
Running phase: buildPhase
@nix { "action": "setPhase", "phase": "buildPhase" }
Executing pypaBuildPhase
Creating a wheel...
* Getting build dependencies for wheel...
running egg_info
creating llm.egg-info
writing llm.egg-info/PKG-INFO
writing dependency_links to llm.egg-info/dependency_links.txt
writing entry points to llm.egg-info/entry_points.txt
writing requirements to llm.egg-info/requires.txt
writing top-level names to llm.egg-info/top_level.txt
writing manifest file 'llm.egg-info/SOURCES.txt'
reading manifest file 'llm.egg-info/SOURCES.txt'
reading manifest template 'MANIFEST.in'
adding license file 'LICENSE'
writing manifest file 'llm.egg-info/SOURCES.txt'
* Building wheel...
running bdist_wheel
running build
running build_py
creating build/lib/llm
copying llm/models.py -> build/lib/llm
copying llm/migrations.py -> build/lib/llm
copying llm/__init__.py -> build/lib/llm
copying llm/cli.py -> build/lib/llm
copying llm/templates.py -> build/lib/llm
copying llm/__main__.py -> build/lib/llm
copying llm/embeddings_migrations.py -> build/lib/llm
copying llm/hookspecs.py -> build/lib/llm
copying llm/utils.py -> build/lib/llm
copying llm/errors.py -> build/lib/llm
copying llm/plugins.py -> build/lib/llm
copying llm/embeddings.py -> build/lib/llm
creating build/lib/llm/default_plugins
copying llm/default_plugins/__init__.py -> build/lib/llm/default_plugins
copying llm/default_plugins/openai_models.py -> build/lib/llm/default_plugins
installing to build/bdist.linux-x86_64/wheel
running install
running install_lib
creating build/bdist.linux-x86_64/wheel
creating build/bdist.linux-x86_64/wheel/llm
copying build/lib/llm/models.py -> build/bdist.linux-x86_64/wheel/./llm
copying build/lib/llm/migrations.py -> build/bdist.linux-x86_64/wheel/./llm
copying build/lib/llm/__init__.py -> build/bdist.linux-x86_64/wheel/./llm
copying build/lib/llm/cli.py -> build/bdist.linux-x86_64/wheel/./llm
copying build/lib/llm/templates.py -> build/bdist.linux-x86_64/wheel/./llm
copying build/lib/llm/__main__.py -> build/bdist.linux-x86_64/wheel/./llm
copying build/lib/llm/embeddings_migrations.py -> build/bdist.linux-x86_64/wheel/./llm
copying build/lib/llm/hookspecs.py -> build/bdist.linux-x86_64/wheel/./llm
copying build/lib/llm/utils.py -> build/bdist.linux-x86_64/wheel/./llm
copying build/lib/llm/errors.py -> build/bdist.linux-x86_64/wheel/./llm
creating build/bdist.linux-x86_64/wheel/llm/default_plugins
copying build/lib/llm/default_plugins/__init__.py -> build/bdist.linux-x86_64/wheel/./llm/default_plugins
copying build/lib/llm/default_plugins/openai_models.py -> build/bdist.linux-x86_64/wheel/./llm/default_plugins
copying build/lib/llm/plugins.py -> build/bdist.linux-x86_64/wheel/./llm
copying build/lib/llm/embeddings.py -> build/bdist.linux-x86_64/wheel/./llm
running install_egg_info
running egg_info
writing llm.egg-info/PKG-INFO
writing dependency_links to llm.egg-info/dependency_links.txt
writing entry points to llm.egg-info/entry_points.txt
writing requirements to llm.egg-info/requires.txt
writing top-level names to llm.egg-info/top_level.txt
reading manifest file 'llm.egg-info/SOURCES.txt'
reading manifest template 'MANIFEST.in'
adding license file 'LICENSE'
writing manifest file 'llm.egg-info/SOURCES.txt'
Copying llm.egg-info to build/bdist.linux-x86_64/wheel/./llm-0.16-py3.12.egg-info
running install_scripts
creating build/bdist.linux-x86_64/wheel/llm-0.16.dist-info/WHEEL
creating '/build/source/dist/.tmp-jgbtcgfa/llm-0.16-py3-none-any.whl' and adding 'build/bdist.linux-x86_64/wheel' to it
adding 'llm/__init__.py'
adding 'llm/__main__.py'
adding 'llm/cli.py'
adding 'llm/embeddings.py'
adding 'llm/embeddings_migrations.py'
adding 'llm/errors.py'
adding 'llm/hookspecs.py'
adding 'llm/migrations.py'
adding 'llm/models.py'
adding 'llm/plugins.py'
adding 'llm/templates.py'
adding 'llm/utils.py'
adding 'llm/default_plugins/__init__.py'
adding 'llm/default_plugins/openai_models.py'
adding 'llm-0.16.dist-info/LICENSE'
adding 'llm-0.16.dist-info/METADATA'
adding 'llm-0.16.dist-info/WHEEL'
adding 'llm-0.16.dist-info/entry_points.txt'
adding 'llm-0.16.dist-info/top_level.txt'
adding 'llm-0.16.dist-info/RECORD'
removing build/bdist.linux-x86_64/wheel
Successfully built llm-0.16-py3-none-any.whl
Finished creating a wheel...
Finished executing pypaBuildPhase
Running phase: pythonRuntimeDepsCheckHook
@nix { "action": "setPhase", "phase": "pythonRuntimeDepsCheckHook" }
Executing pythonRuntimeDepsCheck
Checking runtime dependencies for llm-0.16-py3-none-any.whl
Finished executing pythonRuntimeDepsCheck
Running phase: installPhase
@nix { "action": "setPhase", "phase": "installPhase" }
Executing pypaInstallPhase
Successfully installed llm-0.16-py3-none-any.whl
Finished executing pypaInstallPhase
Running phase: pythonOutputDistPhase
@nix { "action": "setPhase", "phase": "pythonOutputDistPhase" }
Executing pythonOutputDistPhase
Finished executing pythonOutputDistPhase
Running phase: fixupPhase
@nix { "action": "setPhase", "phase": "fixupPhase" }
shrinking RPATHs of ELF executables and libraries in /nix/store/vrwsalqqmmv7drvwdzq80f8qzsr056n0-python3.12-llm-0.16
checking for references to /build/ in /nix/store/vrwsalqqmmv7drvwdzq80f8qzsr056n0-python3.12-llm-0.16...
patching script interpreter paths in /nix/store/vrwsalqqmmv7drvwdzq80f8qzsr056n0-python3.12-llm-0.16
stripping (with command strip and flags -S -p) in  /nix/store/vrwsalqqmmv7drvwdzq80f8qzsr056n0-python3.12-llm-0.16/lib /nix/store/vrwsalqqmmv7drvwdzq80f8qzsr056n0-python3.12-llm-0.16/bin
shrinking RPATHs of ELF executables and libraries in /nix/store/qpjsll2863jar1d5jbwylahh92y34r1j-python3.12-llm-0.16-dist
checking for references to /build/ in /nix/store/qpjsll2863jar1d5jbwylahh92y34r1j-python3.12-llm-0.16-dist...
patching script interpreter paths in /nix/store/qpjsll2863jar1d5jbwylahh92y34r1j-python3.12-llm-0.16-dist
Rewriting #!/nix/store/901c80rlps5q05bnjk1sj4zaz5k736nc-python3-3.12.7/bin/python3.12 to #!/nix/store/901c80rlps5q05bnjk1sj4zaz5k736nc-python3-3.12.7
wrapping `/nix/store/vrwsalqqmmv7drvwdzq80f8qzsr056n0-python3.12-llm-0.16/bin/llm'...
Executing pythonRemoveTestsDir
Finished executing pythonRemoveTestsDir
Running phase: installCheckPhase
@nix { "action": "setPhase", "phase": "installCheckPhase" }
no Makefile or custom installCheckPhase, doing nothing
Running phase: pythonCatchConflictsPhase
@nix { "action": "setPhase", "phase": "pythonCatchConflictsPhase" }
Running phase: pythonRemoveBinBytecodePhase
@nix { "action": "setPhase", "phase": "pythonRemoveBinBytecodePhase" }
Running phase: pythonImportsCheckPhase
@nix { "action": "setPhase", "phase": "pythonImportsCheckPhase" }
Executing pythonImportsCheckPhase
Check whether the following modules can be imported: llm
Running phase: pytestCheckPhase
@nix { "action": "setPhase", "phase": "pytestCheckPhase" }
Executing pytestCheckPhase
============================= test session starts ==============================
platform linux -- Python 3.12.7, pytest-8.3.3, pluggy-1.5.0 -- /nix/store/901c80rlps5q05bnjk1sj4zaz5k736nc-python3-3.12.7/bin/python3.12
cachedir: .pytest_cache
rootdir: /build/source
configfile: pytest.ini
plugins: httpx-0.32.0, anyio-4.6.0
collected 184 items                                                            

tests/test_aliases.py::test_set_alias[gpt-3.5-turbo] PASSED
tests/test_aliases.py::test_set_alias[chatgpt] PASSED
tests/test_aliases.py::test_remove_alias PASSED
tests/test_aliases.py::test_cli_aliases_list[args0] PASSED
tests/test_aliases.py::test_cli_aliases_list[args1] PASSED
tests/test_aliases.py::test_cli_aliases_list_json[args0] PASSED
tests/test_aliases.py::test_cli_aliases_list_json[args1] PASSED
tests/test_aliases.py::test_cli_aliases_set PASSED
tests/test_aliases.py::test_cli_aliases_path PASSED
tests/test_aliases.py::test_cli_aliases_remove PASSED
tests/test_aliases.py::test_cli_aliases_remove_invalid PASSED
tests/test_aliases.py::test_cli_aliases_are_registered[args0] PASSED
tests/test_aliases.py::test_cli_aliases_are_registered[args1] PASSED
tests/test_chat.py::test_mock_model PASSED
tests/test_chat.py::test_chat_basic PASSED
tests/test_chat.py::test_chat_system PASSED
tests/test_chat.py::test_chat_options PASSED
tests/test_chat.py::test_chat_multi[Hi\n!multi\nthis is multiple lines\nuntil the !end\n!end\nquit\n-expected0] PASSED
tests/test_chat.py::test_chat_multi[!multi\nthis is multiple lines\nquit\nuntil the !end\n!end\nquit\n-expected1] PASSED
tests/test_chat.py::test_chat_multi[!multi abc\nCustom delimiter\n!end\n!end 123\n!end abc\nquit\n-expected2] PASSED
tests/test_cli_openai_models.py::test_openai_models PASSED
tests/test_cli_openai_models.py::test_openai_options_min_max PASSED
tests/test_embed.py::test_demo_plugin PASSED
tests/test_embed.py::test_embed_huge_list[None-100] PASSED
tests/test_embed.py::test_embed_huge_list[10-100] PASSED
tests/test_embed.py::test_embed_store PASSED
tests/test_embed.py::test_embed_metadata PASSED
tests/test_embed.py::test_collection PASSED
tests/test_embed.py::test_similar PASSED
tests/test_embed.py::test_similar_by_id PASSED
tests/test_embed.py::test_embed_multi[False-None-100] PASSED
tests/test_embed.py::test_embed_multi[False-5-200] PASSED
tests/test_embed.py::test_embed_multi[True-None-100] PASSED
tests/test_embed.py::test_embed_multi[True-5-200] PASSED
tests/test_embed.py::test_collection_delete PASSED
tests/test_embed.py::test_binary_only_and_text_only_embedding_models PASSED
tests/test_embed_cli.py::test_embed_output_format[argument-json-[5, 5, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0]\n] PASSED
tests/test_embed_cli.py::test_embed_output_format[argument-base64-AACgQAAAoEAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAA==\n] PASSED
tests/test_embed_cli.py::test_embed_output_format[argument-hex-0000a0400000a0400000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000\n] PASSED
tests/test_embed_cli.py::test_embed_output_format[argument-blob-\x00\x00\ufffd@\x00\x00\ufffd@\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\n] PASSED
tests/test_embed_cli.py::test_embed_output_format[file-json-[5, 5, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0]\n] PASSED
tests/test_embed_cli.py::test_embed_output_format[file-base64-AACgQAAAoEAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAA==\n] PASSED
tests/test_embed_cli.py::test_embed_output_format[file-hex-0000a0400000a0400000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000\n] PASSED
tests/test_embed_cli.py::test_embed_output_format[file-blob-\x00\x00\ufffd@\x00\x00\ufffd@\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\n] PASSED
tests/test_embed_cli.py::test_embed_output_format[stdin-json-[5, 5, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0]\n] PASSED
tests/test_embed_cli.py::test_embed_output_format[stdin-base64-AACgQAAAoEAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAA==\n] PASSED
tests/test_embed_cli.py::test_embed_output_format[stdin-hex-0000a0400000a0400000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000\n] PASSED
tests/test_embed_cli.py::test_embed_output_format[stdin-blob-\x00\x00\ufffd@\x00\x00\ufffd@\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\n] PASSED
tests/test_embed_cli.py::test_embed_errors[args0-Must provide both collection and id] PASSED
tests/test_embed_cli.py::test_embed_store[None-None] PASSED
tests/test_embed_cli.py::test_embed_store[{"foo": "bar"}-None] PASSED
tests/test_embed_cli.py::test_embed_store[{"foo": [1, 2, 3]}-None] PASSED
tests/test_embed_cli.py::test_embed_store[[1, 2, 3]-Metadata must be a JSON object] PASSED
tests/test_embed_cli.py::test_embed_store[{"foo": "incomplete}-Metadata must be valid JSON] PASSED
tests/test_embed_cli.py::test_embed_store_binary PASSED
tests/test_embed_cli.py::test_collection_delete_errors PASSED
tests/test_embed_cli.py::test_similar_errors[args0-Missing argument 'COLLECTION'] PASSED
tests/test_embed_cli.py::test_similar_errors[args1-Collection does not exist] PASSED
tests/test_embed_cli.py::test_similar_errors[args2-ID not found in collection] PASSED
tests/test_embed_cli.py::test_similar_by_id_cli PASSED
tests/test_embed_cli.py::test_similar_by_content_cli[argument] PASSED
tests/test_embed_cli.py::test_similar_by_content_cli[file] PASSED
tests/test_embed_cli.py::test_similar_by_content_cli[stdin] PASSED
tests/test_embed_cli.py::test_embed_multi_file_input[phrases.csv-id,phrase\n1,hello world\n2,goodbye world-None-False] PASSED
tests/test_embed_cli.py::test_embed_multi_file_input[phrases.csv-id,phrase\n1,hello world\n2,goodbye world-None-True] PASSED
tests/test_embed_cli.py::test_embed_multi_file_input[phrases.csv-id,phrase\n1,hello world\n2,goodbye world-prefix-False] PASSED
tests/test_embed_cli.py::test_embed_multi_file_input[phrases.csv-id,phrase\n1,hello world\n2,goodbye world-prefix-True] PASSED
tests/test_embed_cli.py::test_embed_multi_file_input[phrases.tsv-id\tphrase\n1\thello world\n2\tgoodbye world-None-False] PASSED
tests/test_embed_cli.py::test_embed_multi_file_input[phrases.tsv-id\tphrase\n1\thello world\n2\tgoodbye world-None-True] PASSED
tests/test_embed_cli.py::test_embed_multi_file_input[phrases.tsv-id\tphrase\n1\thello world\n2\tgoodbye world-prefix-False] PASSED
tests/test_embed_cli.py::test_embed_multi_file_input[phrases.tsv-id\tphrase\n1\thello world\n2\tgoodbye world-prefix-True] PASSED
tests/test_embed_cli.py::test_embed_multi_file_input[phrases.jsonl-{"id": 1, "phrase": "hello world"}\n{"id": 2, "phrase": "goodbye world"}-None-False] PASSED
tests/test_embed_cli.py::test_embed_multi_file_input[phrases.jsonl-{"id": 1, "phrase": "hello world"}\n{"id": 2, "phrase": "goodbye world"}-None-True] PASSED
tests/test_embed_cli.py::test_embed_multi_file_input[phrases.jsonl-{"id": 1, "phrase": "hello world"}\n{"id": 2, "phrase": "goodbye world"}-prefix-False] PASSED
tests/test_embed_cli.py::test_embed_multi_file_input[phrases.jsonl-{"id": 1, "phrase": "hello world"}\n{"id": 2, "phrase": "goodbye world"}-prefix-True] PASSED
tests/test_embed_cli.py::test_embed_multi_file_input[phrases.json-[{"id": 1, "phrase": "hello world"}, {"id": 2, "phrase": "goodbye world"}]-None-False] PASSED
tests/test_embed_cli.py::test_embed_multi_file_input[phrases.json-[{"id": 1, "phrase": "hello world"}, {"id": 2, "phrase": "goodbye world"}]-None-True] PASSED
tests/test_embed_cli.py::test_embed_multi_file_input[phrases.json-[{"id": 1, "phrase": "hello world"}, {"id": 2, "phrase": "goodbye world"}]-prefix-False] PASSED
tests/test_embed_cli.py::test_embed_multi_file_input[phrases.json-[{"id": 1, "phrase": "hello world"}, {"id": 2, "phrase": "goodbye world"}]-prefix-True] PASSED
tests/test_embed_cli.py::test_embed_multi_files_binary_store PASSED
tests/test_embed_cli.py::test_embed_multi_sql[None-True] PASSED
tests/test_embed_cli.py::test_embed_multi_sql[None-False] PASSED
tests/test_embed_cli.py::test_embed_multi_sql[prefix-True] PASSED
tests/test_embed_cli.py::test_embed_multi_sql[prefix-False] PASSED
tests/test_embed_cli.py::test_embed_multi_batch_size PASSED
tests/test_embed_cli.py::test_embed_multi_files[single] PASSED
tests/test_embed_cli.py::test_embed_multi_files[multi] PASSED
tests/test_embed_cli.py::test_embed_multi_files_errors[args0-Invalid directory: not-a-dir] PASSED
tests/test_embed_cli.py::test_embed_multi_files_encoding[extra_args0-None] PASSED
tests/test_embed_cli.py::test_embed_multi_files_encoding[extra_args1-Could not decode text in file] PASSED
tests/test_embed_cli.py::test_embed_multi_files_encoding[extra_args2-None] PASSED
tests/test_embed_cli.py::test_embed_multi_files_encoding[extra_args3-None] PASSED
tests/test_embed_cli.py::test_embed_multi_files_encoding[extra_args4-None] PASSED
tests/test_embed_cli.py::test_default_embedding_model PASSED
tests/test_embed_cli.py::test_default_embed_model_errors[embed-False] PASSED
tests/test_embed_cli.py::test_default_embed_model_errors[embed-True] PASSED
tests/test_embed_cli.py::test_default_embed_model_errors[embed-multi-False] PASSED
tests/test_embed_cli.py::test_default_embed_model_errors[embed-multi-True] PASSED
tests/test_embed_cli.py::test_duplicate_content_embedded_only_once PASSED
tests/test_encode_decode.py::test_roundtrip[array0] PASSED
tests/test_encode_decode.py::test_roundtrip[array1] PASSED
tests/test_keys.py::test_keys_in_user_path[env0] PASSED
tests/test_keys.py::test_keys_in_user_path[env1] PASSED
tests/test_keys.py::test_keys_set PASSED
tests/test_keys.py::test_keys_list[args0] PASSED
tests/test_keys.py::test_keys_list[args1] PASSED
tests/test_keys.py::test_uses_correct_key FAILED
tests/test_keys.py::test_uses_correct_key ERROR
tests/test_llm.py::test_version PASSED
tests/test_llm.py::test_logs_text PASSED
tests/test_llm.py::test_logs_json[None] PASSED
tests/test_llm.py::test_logs_json[0] PASSED
tests/test_llm.py::test_logs_json[2] PASSED
tests/test_llm.py::test_logs_response_only[args0] PASSED
tests/test_llm.py::test_logs_response_only[args1] PASSED
tests/test_llm.py::test_logs_response_only[args2] PASSED
tests/test_llm.py::test_logs_response_only[args3] PASSED
tests/test_llm.py::test_logs_path[env0] PASSED
tests/test_llm.py::test_logs_path[env1] PASSED
tests/test_llm.py::test_logs_filtered[davinci] PASSED
tests/test_llm.py::test_logs_filtered[curie] PASSED
tests/test_llm.py::test_logs_search[-extra_args0-expected0] PASSED
tests/test_llm.py::test_logs_search[llama-extra_args1-expected1] PASSED
tests/test_llm.py::test_logs_search[alpaca-extra_args2-expected2] PASSED
tests/test_llm.py::test_logs_search[llama-extra_args3-expected3] PASSED
tests/test_llm.py::test_logs_search[llama-extra_args4-expected4] PASSED
tests/test_llm.py::test_llm_prompt_creates_log_database PASSED
tests/test_llm.py::test_llm_default_prompt[True-logs_args0-False-True] PASSED
tests/test_llm.py::test_llm_default_prompt[True-logs_args0-False-False] PASSED
tests/test_llm.py::test_llm_default_prompt[True-logs_args0-False-split] PASSED
tests/test_llm.py::test_llm_default_prompt[False-logs_args1-True-True] PASSED
tests/test_llm.py::test_llm_default_prompt[False-logs_args1-True-False] PASSED
tests/test_llm.py::test_llm_default_prompt[False-logs_args1-True-split] PASSED
tests/test_llm.py::test_llm_default_prompt[False-logs_args2-False-True] PASSED
tests/test_llm.py::test_llm_default_prompt[False-logs_args2-False-False] PASSED
tests/test_llm.py::test_llm_default_prompt[False-logs_args2-False-split] PASSED
tests/test_llm.py::test_llm_default_prompt[False-logs_args3-True-True] PASSED
tests/test_llm.py::test_llm_default_prompt[False-logs_args3-True-False] PASSED
tests/test_llm.py::test_llm_default_prompt[False-logs_args3-True-split] PASSED
tests/test_llm.py::test_llm_default_prompt[True-logs_args4-False-True] PASSED
tests/test_llm.py::test_llm_default_prompt[True-logs_args4-False-False] PASSED
tests/test_llm.py::test_llm_default_prompt[True-logs_args4-False-split] PASSED
tests/test_llm.py::test_llm_default_prompt[True-logs_args5-True-True] PASSED
tests/test_llm.py::test_llm_default_prompt[True-logs_args5-True-False] PASSED
tests/test_llm.py::test_llm_default_prompt[True-logs_args5-True-split] PASSED
tests/test_llm.py::test_openai_chat_stream PASSED
tests/test_llm.py::test_openai_completion PASSED
tests/test_llm.py::test_openai_completion_system_prompt_error PASSED
tests/test_llm.py::test_openai_completion_logprobs_stream PASSED
tests/test_llm.py::test_openai_completion_logprobs_nostream PASSED
tests/test_llm.py::test_openai_localai_configuration PASSED
tests/test_llm.py::test_llm_models_options PASSED
tests/test_llm.py::test_llm_user_dir PASSED
tests/test_llm.py::test_model_defaults PASSED
tests/test_migrate.py::test_migrate_blank PASSED
tests/test_migrate.py::test_migrate_from_original_schema[True] PASSED
tests/test_migrate.py::test_migrate_from_original_schema[False] PASSED
tests/test_migrate.py::test_migrations_with_legacy_alter_table PASSED
tests/test_migrate.py::test_migrations_for_embeddings PASSED
tests/test_migrate.py::test_backfill_content_hash PASSED
tests/test_plugins.py::test_register_commands PASSED
tests/test_templates.py::test_template_evaluate[S: $input-None-None-params0-S: input-None-None] PASSED
tests/test_templates.py::test_template_evaluate[S: $input-system-None-params1-S: input-system-None] PASSED
tests/test_templates.py::test_template_evaluate[No vars-None-None-params2-No vars-None-None] PASSED
tests/test_templates.py::test_template_evaluate[$one and $two-None-None-params3-None-None-Missing variables: one, two] PASSED
tests/test_templates.py::test_template_evaluate[$one and $two-None-None-params4-1 and 2-None-None] PASSED
tests/test_templates.py::test_template_evaluate[$one and $two-None-defaults5-params5-1 and 2-None-None] PASSED
tests/test_templates.py::test_template_evaluate[$one and $two-None-defaults6-params6-1 and 2-None-None] PASSED
tests/test_templates.py::test_templates_list_no_templates_found PASSED
tests/test_templates.py::test_templates_list[args0] PASSED
tests/test_templates.py::test_templates_list[args1] PASSED
tests/test_templates.py::test_templates_prompt_save[args0-expected_prompt0-None] PASSED
tests/test_templates.py::test_templates_prompt_save[args1-expected_prompt1-None] PASSED
tests/test_templates.py::test_templates_prompt_save[args2-expected_prompt2-None] PASSED
tests/test_templates.py::test_templates_prompt_save[args3-None---save cannot be used with --template] PASSED
tests/test_templates.py::test_templates_prompt_save[args4-None---save cannot be used with --continue] PASSED
tests/test_templates.py::test_templates_prompt_save[args5-None---save cannot be used with --cid] PASSED
tests/test_templates.py::test_templates_prompt_save[args6-None---save cannot be used with --cid] PASSED
tests/test_templates.py::test_templates_prompt_save[args7-expected_prompt7-None] PASSED
tests/test_templates.py::test_template_basic['Summarize this: $input'-extra_args0-gpt-4o-mini-Summarize this: Input text-None] PASSED
tests/test_templates.py::test_template_basic[prompt: 'Summarize this: $input'\nmodel: gpt-4-extra_args1-gpt-4-Summarize this: Input text-None] PASSED
tests/test_templates.py::test_template_basic[prompt: 'Summarize this: $input'-extra_args2-gpt-4-Summarize this: Input text-None] PASSED
tests/test_templates.py::test_template_basic[boo-extra_args3-None-None-Error: Cannot use -t/--template and --system together] FAILED
tests/test_templates.py::test_template_basic[boo-extra_args3-None-None-Error: Cannot use -t/--template and --system together] ERROR
tests/test_templates.py::test_template_basic[prompt: 'Say $hello'-extra_args4-None-None-Error: Missing variables: hello] FAILED
tests/test_templates.py::test_template_basic[prompt: 'Say $hello'-extra_args4-None-None-Error: Missing variables: hello] ERROR
tests/test_templates.py::test_template_basic[prompt: 'Say $hello'-extra_args5-gpt-4o-mini-Say Blah-None] PASSED

==================================== ERRORS ====================================
__________________ ERROR at teardown of test_uses_correct_key __________________

monkeypatch = <_pytest.monkeypatch.MonkeyPatch object at 0x7fff71986ab0>
request = <SubRequest 'httpx_mock' for <Function test_uses_correct_key>>

    @pytest.fixture
    def httpx_mock(
        monkeypatch: MonkeyPatch,
        request: FixtureRequest,
    ) -> Generator[HTTPXMock, None, None]:
        options = {}
        for marker in request.node.iter_markers("httpx_mock"):
            options = marker.kwargs | options
        __tracebackhide__ = methodcaller("errisinstance", TypeError)
        options = _HTTPXMockOptions(**options)
    
        mock = HTTPXMock(options)
    
        # Mock synchronous requests
        real_handle_request = httpx.HTTPTransport.handle_request
    
        def mocked_handle_request(
            transport: httpx.HTTPTransport, request: httpx.Request
        ) -> httpx.Response:
            if request.url.host in options.non_mocked_hosts:
                return real_handle_request(transport, request)
            return mock._handle_request(transport, request)
    
        monkeypatch.setattr(
            httpx.HTTPTransport,
            "handle_request",
            mocked_handle_request,
        )
    
        # Mock asynchronous requests
        real_handle_async_request = httpx.AsyncHTTPTransport.handle_async_request
    
        async def mocked_handle_async_request(
            transport: httpx.AsyncHTTPTransport, request: httpx.Request
        ) -> httpx.Response:
            if request.url.host in options.non_mocked_hosts:
                return await real_handle_async_request(transport, request)
            return await mock._handle_async_request(transport, request)
    
        monkeypatch.setattr(
            httpx.AsyncHTTPTransport,
            "handle_async_request",
            mocked_handle_async_request,
        )
    
        yield mock
        try:
>           mock._assert_options()

/nix/store/1nzsgd8b21jv6jqas84502lw226ygbgl-python3.12-pytest-httpx-0.32.0/lib/python3.12/site-packages/pytest_httpx/__init__.py:67: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

self = <pytest_httpx._httpx_mock.HTTPXMock object at 0x7fff7179c3b0>

    def _assert_options(self) -> None:
        if self._options.assert_all_responses_were_requested:
            callbacks_not_executed = [
                matcher for matcher, _ in self._callbacks if not matcher.nb_calls
            ]
            matchers_description = "\n".join(
                [f"- {matcher}" for matcher in callbacks_not_executed]
            )
    
            assert not callbacks_not_executed, (
                "The following responses are mocked but not requested:\n"
                f"{matchers_description}\n"
                "\n"
                "If this is on purpose, refer to https://github.com/Colin-b/pytest_httpx/blob/master/README.md#allow-to-register-more-responses-than-what-will-be-requested"
            )
    
        if self._options.assert_all_requests_were_expected:
            requests_description = "\n".join(
                [
                    f"- {request.method} request on {request.url}"
                    for request in self._requests_not_matched
                ]
            )
>           assert not self._requests_not_matched, (
                f"The following requests were not expected:\n"
                f"{requests_description}\n"
                "\n"
                "If this is on purpose, refer to https://github.com/Colin-b/pytest_httpx/blob/master/README.md#allow-to-not-register-responses-for-every-request"
            )
E           AssertionError: The following requests were not expected:
E             - POST request on https://api.openai.com/v1/chat/completions
E             - POST request on https://api.openai.com/v1/chat/completions
E             - POST request on https://api.openai.com/v1/chat/completions
E             
E             If this is on purpose, refer to https://github.com/Colin-b/pytest_httpx/blob/master/README.md#allow-to-not-register-responses-for-every-request
E           assert not [<Request('POST', 'https://api.openai.com/v1/chat/completions')>, <Request('POST', 'https://api.openai.com/v1/chat/completions')>, <Request('POST', 'https://api.openai.com/v1/chat/completions')>]
E            +  where [<Request('POST', 'https://api.openai.com/v1/chat/completions')>, <Request('POST', 'https://api.openai.com/v1/chat/completions')>, <Request('POST', 'https://api.openai.com/v1/chat/completions')>] = <pytest_httpx._httpx_mock.HTTPXMock object at 0x7fff7179c3b0>._requests_not_matched

/nix/store/1nzsgd8b21jv6jqas84502lw226ygbgl-python3.12-pytest-httpx-0.32.0/lib/python3.12/site-packages/pytest_httpx/_httpx_mock.py:314: AssertionError
_ ERROR at teardown of test_template_basic[boo-extra_args3-None-None-Error: Cannot use -t/--template and --system together] _

monkeypatch = <_pytest.monkeypatch.MonkeyPatch object at 0x7fff71620b00>
request = <SubRequest 'httpx_mock' for <Function test_template_basic[boo-extra_args3-None-None-Error: Cannot use -t/--template and --system together]>>

    @pytest.fixture
    def httpx_mock(
        monkeypatch: MonkeyPatch,
        request: FixtureRequest,
    ) -> Generator[HTTPXMock, None, None]:
        options = {}
        for marker in request.node.iter_markers("httpx_mock"):
            options = marker.kwargs | options
        __tracebackhide__ = methodcaller("errisinstance", TypeError)
        options = _HTTPXMockOptions(**options)
    
        mock = HTTPXMock(options)
    
        # Mock synchronous requests
        real_handle_request = httpx.HTTPTransport.handle_request
    
        def mocked_handle_request(
            transport: httpx.HTTPTransport, request: httpx.Request
        ) -> httpx.Response:
            if request.url.host in options.non_mocked_hosts:
                return real_handle_request(transport, request)
            return mock._handle_request(transport, request)
    
        monkeypatch.setattr(
            httpx.HTTPTransport,
            "handle_request",
            mocked_handle_request,
        )
    
        # Mock asynchronous requests
        real_handle_async_request = httpx.AsyncHTTPTransport.handle_async_request
    
        async def mocked_handle_async_request(
            transport: httpx.AsyncHTTPTransport, request: httpx.Request
        ) -> httpx.Response:
            if request.url.host in options.non_mocked_hosts:
                return await real_handle_async_request(transport, request)
            return await mock._handle_async_request(transport, request)
    
        monkeypatch.setattr(
            httpx.AsyncHTTPTransport,
            "handle_async_request",
            mocked_handle_async_request,
        )
    
        yield mock
        try:
>           mock._assert_options()

/nix/store/1nzsgd8b21jv6jqas84502lw226ygbgl-python3.12-pytest-httpx-0.32.0/lib/python3.12/site-packages/pytest_httpx/__init__.py:67: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

self = <pytest_httpx._httpx_mock.HTTPXMock object at 0x7fff7168a9c0>

    def _assert_options(self) -> None:
        if self._options.assert_all_responses_were_requested:
            callbacks_not_executed = [
                matcher for matcher, _ in self._callbacks if not matcher.nb_calls
            ]
            matchers_description = "\n".join(
                [f"- {matcher}" for matcher in callbacks_not_executed]
            )
    
>           assert not callbacks_not_executed, (
                "The following responses are mocked but not requested:\n"
                f"{matchers_description}\n"
                "\n"
                "If this is on purpose, refer to https://github.com/Colin-b/pytest_httpx/blob/master/README.md#allow-to-register-more-responses-than-what-will-be-requested"
            )
E           AssertionError: The following responses are mocked but not requested:
E             - Match POST request on https://api.openai.com/v1/chat/completions
E             
E             If this is on purpose, refer to https://github.com/Colin-b/pytest_httpx/blob/master/README.md#allow-to-register-more-responses-than-what-will-be-requested
E           assert not [<pytest_httpx._request_matcher._RequestMatcher object at 0x7fff7160f0e0>]

/nix/store/1nzsgd8b21jv6jqas84502lw226ygbgl-python3.12-pytest-httpx-0.32.0/lib/python3.12/site-packages/pytest_httpx/_httpx_mock.py:300: AssertionError
_ ERROR at teardown of test_template_basic[prompt: 'Say $hello'-extra_args4-None-None-Error: Missing variables: hello] _

monkeypatch = <_pytest.monkeypatch.MonkeyPatch object at 0x7fff71497740>
request = <SubRequest 'httpx_mock' for <Function test_template_basic[prompt: 'Say $hello'-extra_args4-None-None-Error: Missing variables: hello]>>

    @pytest.fixture
    def httpx_mock(
        monkeypatch: MonkeyPatch,
        request: FixtureRequest,
    ) -> Generator[HTTPXMock, None, None]:
        options = {}
        for marker in request.node.iter_markers("httpx_mock"):
            options = marker.kwargs | options
        __tracebackhide__ = methodcaller("errisinstance", TypeError)
        options = _HTTPXMockOptions(**options)
    
        mock = HTTPXMock(options)
    
        # Mock synchronous requests
        real_handle_request = httpx.HTTPTransport.handle_request
    
        def mocked_handle_request(
            transport: httpx.HTTPTransport, request: httpx.Request
        ) -> httpx.Response:
            if request.url.host in options.non_mocked_hosts:
                return real_handle_request(transport, request)
            return mock._handle_request(transport, request)
    
        monkeypatch.setattr(
            httpx.HTTPTransport,
            "handle_request",
            mocked_handle_request,
        )
    
        # Mock asynchronous requests
        real_handle_async_request = httpx.AsyncHTTPTransport.handle_async_request
    
        async def mocked_handle_async_request(
            transport: httpx.AsyncHTTPTransport, request: httpx.Request
        ) -> httpx.Response:
            if request.url.host in options.non_mocked_hosts:
                return await real_handle_async_request(transport, request)
            return await mock._handle_async_request(transport, request)
    
        monkeypatch.setattr(
            httpx.AsyncHTTPTransport,
            "handle_async_request",
            mocked_handle_async_request,
        )
    
        yield mock
        try:
>           mock._assert_options()

/nix/store/1nzsgd8b21jv6jqas84502lw226ygbgl-python3.12-pytest-httpx-0.32.0/lib/python3.12/site-packages/pytest_httpx/__init__.py:67: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

self = <pytest_httpx._httpx_mock.HTTPXMock object at 0x7fff71497140>

    def _assert_options(self) -> None:
        if self._options.assert_all_responses_were_requested:
            callbacks_not_executed = [
                matcher for matcher, _ in self._callbacks if not matcher.nb_calls
            ]
            matchers_description = "\n".join(
                [f"- {matcher}" for matcher in callbacks_not_executed]
            )
    
>           assert not callbacks_not_executed, (
                "The following responses are mocked but not requested:\n"
                f"{matchers_description}\n"
                "\n"
                "If this is on purpose, refer to https://github.com/Colin-b/pytest_httpx/blob/master/README.md#allow-to-register-more-responses-than-what-will-be-requested"
            )
E           AssertionError: The following responses are mocked but not requested:
E             - Match POST request on https://api.openai.com/v1/chat/completions
E             
E             If this is on purpose, refer to https://github.com/Colin-b/pytest_httpx/blob/master/README.md#allow-to-register-more-responses-than-what-will-be-requested
E           assert not [<pytest_httpx._request_matcher._RequestMatcher object at 0x7fff71496a50>]

/nix/store/1nzsgd8b21jv6jqas84502lw226ygbgl-python3.12-pytest-httpx-0.32.0/lib/python3.12/site-packages/pytest_httpx/_httpx_mock.py:300: AssertionError
=================================== FAILURES ===================================
____________________________ test_uses_correct_key _____________________________

mocked_openai_chat = <pytest_httpx._httpx_mock.HTTPXMock object at 0x7fff7179c3b0>
monkeypatch = <_pytest.monkeypatch.MonkeyPatch object at 0x7fff71986ab0>
tmpdir = local('/build/pytest-of-nixbld/pytest-0/test_uses_correct_key0')

    def test_uses_correct_key(mocked_openai_chat, monkeypatch, tmpdir):
        user_dir = tmpdir / "user-dir"
        pathlib.Path(user_dir).mkdir()
        keys_path = user_dir / "keys.json"
        KEYS = {
            "openai": "from-keys-file",
            "other": "other-key",
        }
        keys_path.write_text(json.dumps(KEYS), "utf-8")
        monkeypatch.setenv("LLM_USER_PATH", str(user_dir))
        monkeypatch.setenv("OPENAI_API_KEY", "from-env")
    
        def assert_key(key):
            request = mocked_openai_chat.get_requests()[-1]
            assert request.headers["Authorization"] == "Bearer {}".format(key)
    
        runner = CliRunner()
    
        # Called without --key uses stored key
        result = runner.invoke(cli, ["hello", "--no-stream"], catch_exceptions=False)
        assert result.exit_code == 0
        assert_key("from-keys-file")
    
        # Called without --key and without keys.json uses environment variable
        keys_path.write_text("{}", "utf-8")
        result2 = runner.invoke(cli, ["hello", "--no-stream"], catch_exceptions=False)
>       assert result2.exit_code == 0
E       assert 1 == 0
E        +  where 1 = <Result SystemExit(1)>.exit_code

tests/test_keys.py:81: AssertionError
_ test_template_basic[boo-extra_args3-None-None-Error: Cannot use -t/--template and --system together] _

templates_path = local('/build/pytest-of-nixbld/pytest-0/test_template_basic_boo_extra_0/llm.datasette.io/templates')
mocked_openai_chat = <pytest_httpx._httpx_mock.HTTPXMock object at 0x7fff7168a9c0>
template = 'boo', extra_args = ['-s', 's'], expected_model = None
expected_input = None
expected_error = 'Error: Cannot use -t/--template and --system together'

    @mock.patch.dict(os.environ, {"OPENAI_API_KEY": "X"})
    @pytest.mark.parametrize(
        "template,extra_args,expected_model,expected_input,expected_error",
        (
            (
                "'Summarize this: $input'",
                [],
                "gpt-4o-mini",
                "Summarize this: Input text",
                None,
            ),
            (
                "prompt: 'Summarize this: $input'\nmodel: gpt-4",
                [],
                "gpt-4",
                "Summarize this: Input text",
                None,
            ),
            (
                "prompt: 'Summarize this: $input'",
                ["-m", "4"],
                "gpt-4",
                "Summarize this: Input text",
                None,
            ),
            (
                "boo",
                ["-s", "s"],
                None,
                None,
                "Error: Cannot use -t/--template and --system together",
            ),
            (
                "prompt: 'Say $hello'",
                [],
                None,
                None,
                "Error: Missing variables: hello",
            ),
            (
                "prompt: 'Say $hello'",
                ["-p", "hello", "Blah"],
                "gpt-4o-mini",
                "Say Blah",
                None,
            ),
        ),
    )
    def test_template_basic(
        templates_path,
        mocked_openai_chat,
        template,
        extra_args,
        expected_model,
        expected_input,
        expected_error,
    ):
        (templates_path / "template.yaml").write_text(template, "utf-8")
        runner = CliRunner()
        result = runner.invoke(
            cli,
            ["--no-stream", "-t", "template", "Input text"] + extra_args,
            catch_exceptions=False,
        )
        if expected_error is None:
            assert result.exit_code == 0
            last_request = mocked_openai_chat.get_requests()[-1]
            assert json.loads(last_request.content) == {
                "model": expected_model,
                "messages": [{"role": "user", "content": expected_input}],
                "stream": False,
            }
        else:
            assert result.exit_code == 1
            assert result.output.strip() == expected_error
>           mocked_openai_chat.reset(assert_all_responses_were_requested=False)
E           TypeError: HTTPXMock.reset() got an unexpected keyword argument 'assert_all_responses_were_requested'

tests/test_templates.py:186: TypeError
_ test_template_basic[prompt: 'Say $hello'-extra_args4-None-None-Error: Missing variables: hello] _

templates_path = local('/build/pytest-of-nixbld/pytest-0/test_template_basic_prompt___S2/llm.datasette.io/templates')
mocked_openai_chat = <pytest_httpx._httpx_mock.HTTPXMock object at 0x7fff71497140>
template = "prompt: 'Say $hello'", extra_args = [], expected_model = None
expected_input = None, expected_error = 'Error: Missing variables: hello'

    @mock.patch.dict(os.environ, {"OPENAI_API_KEY": "X"})
    @pytest.mark.parametrize(
        "template,extra_args,expected_model,expected_input,expected_error",
        (
            (
                "'Summarize this: $input'",
                [],
                "gpt-4o-mini",
                "Summarize this: Input text",
                None,
            ),
            (
                "prompt: 'Summarize this: $input'\nmodel: gpt-4",
                [],
                "gpt-4",
                "Summarize this: Input text",
                None,
            ),
            (
                "prompt: 'Summarize this: $input'",
                ["-m", "4"],
                "gpt-4",
                "Summarize this: Input text",
                None,
            ),
            (
                "boo",
                ["-s", "s"],
                None,
                None,
                "Error: Cannot use -t/--template and --system together",
            ),
            (
                "prompt: 'Say $hello'",
                [],
                None,
                None,
                "Error: Missing variables: hello",
            ),
            (
                "prompt: 'Say $hello'",
                ["-p", "hello", "Blah"],
                "gpt-4o-mini",
                "Say Blah",
                None,
            ),
        ),
    )
    def test_template_basic(
        templates_path,
        mocked_openai_chat,
        template,
        extra_args,
        expected_model,
        expected_input,
        expected_error,
    ):
        (templates_path / "template.yaml").write_text(template, "utf-8")
        runner = CliRunner()
        result = runner.invoke(
            cli,
            ["--no-stream", "-t", "template", "Input text"] + extra_args,
            catch_exceptions=False,
        )
        if expected_error is None:
            assert result.exit_code == 0
            last_request = mocked_openai_chat.get_requests()[-1]
            assert json.loads(last_request.content) == {
                "model": expected_model,
                "messages": [{"role": "user", "content": expected_input}],
                "stream": False,
            }
        else:
            assert result.exit_code == 1
            assert result.output.strip() == expected_error
>           mocked_openai_chat.reset(assert_all_responses_were_requested=False)
E           TypeError: HTTPXMock.reset() got an unexpected keyword argument 'assert_all_responses_were_requested'

tests/test_templates.py:186: TypeError
=============================== warnings summary ===============================
tests/test_chat.py: 11 warnings
tests/test_keys.py: 2 warnings
tests/test_llm.py: 26 warnings
tests/test_templates.py: 4 warnings
  /build/source/llm/models.py:88: DeprecationWarning: datetime.datetime.utcnow() is deprecated and scheduled for removal in a future version. Use timezone-aware objects to represent datetimes in UTC: datetime.datetime.now(datetime.UTC).
    self._start_utcnow = datetime.datetime.utcnow()

tests/test_chat.py: 78 warnings
tests/test_keys.py: 13 warnings
tests/test_llm.py: 390 warnings
tests/test_migrate.py: 52 warnings
tests/test_templates.py: 52 warnings
  /build/source/llm/migrations.py:16: DeprecationWarning: datetime.datetime.utcnow() is deprecated and scheduled for removal in a future version. Use timezone-aware objects to represent datetimes in UTC: datetime.datetime.now(datetime.UTC).
    {"name": name, "applied_at": str(datetime.datetime.utcnow())}

tests/test_cli_openai_models.py::test_openai_models
tests/test_cli_openai_models.py::test_openai_models
  /build/source/llm/default_plugins/openai_models.py:149: DeprecationWarning: datetime.datetime.utcfromtimestamp() is deprecated and scheduled for removal in a future version. Use timezone-aware objects to represent datetimes in UTC: datetime.datetime.fromtimestamp(timestamp, datetime.UTC).
    created_str = datetime.datetime.utcfromtimestamp(

tests/test_embed.py: 50 warnings
tests/test_embed_cli.py: 235 warnings
tests/test_migrate.py: 10 warnings
  /nix/store/7jni9vmab6qyspz4lc6hpva6v5656qdl-python3.12-sqlite-migrate-0.1a2/lib/python3.12/site-packages/sqlite_migrate/__init__.py:80: DeprecationWarning: datetime.datetime.utcnow() is deprecated and scheduled for removal in a future version. Use timezone-aware objects to represent datetimes in UTC: datetime.datetime.now(datetime.UTC).
    "applied_at": str(datetime.datetime.utcnow()),

tests/test_llm.py::test_logs_text
tests/test_llm.py::test_logs_json[None]
tests/test_llm.py::test_logs_json[0]
tests/test_llm.py::test_logs_json[2]
tests/test_llm.py::test_logs_response_only[args0]
tests/test_llm.py::test_logs_response_only[args1]
tests/test_llm.py::test_logs_response_only[args2]
tests/test_llm.py::test_logs_response_only[args3]
  /build/source/tests/test_llm.py:30: DeprecationWarning: datetime.datetime.utcnow() is deprecated and scheduled for removal in a future version. Use timezone-aware objects to represent datetimes in UTC: datetime.datetime.now(datetime.UTC).
    start = datetime.datetime.utcnow()

-- Docs: https://docs.pytest.org/en/stable/how-to/capture-warnings.html
=========================== short test summary info ============================
FAILED tests/test_keys.py::test_uses_correct_key - assert 1 == 0
 +  where 1 = <Result SystemExit(1)>.exit_code
FAILED tests/test_templates.py::test_template_basic[boo-extra_args3-None-None-Error: Cannot use -t/--template and --system together] - TypeError: HTTPXMock.reset() got an unexpected keyword argument 'assert_all_responses_were_requested'
FAILED tests/test_templates.py::test_template_basic[prompt: 'Say $hello'-extra_args4-None-None-Error: Missing variables: hello] - TypeError: HTTPXMock.reset() got an unexpected keyword argument 'assert_all_responses_were_requested'
ERROR tests/test_keys.py::test_uses_correct_key - AssertionError: The following requests were not expected:
  - POST request on https://api.openai.com/v1/chat/completions
  - POST request on https://api.openai.com/v1/chat/completions
  - POST request on https://api.openai.com/v1/chat/completions
  
  If this is on purpose, refer to https://github.com/Colin-b/pytest_httpx/blob/master/README.md#allow-to-not-register-responses-for-every-request
assert not [<Request('POST', 'https://api.openai.com/v1/chat/completions')>, <Request('POST', 'https://api.openai.com/v1/chat/completions')>, <Request('POST', 'https://api.openai.com/v1/chat/completions')>]
 +  where [<Request('POST', 'https://api.openai.com/v1/chat/completions')>, <Request('POST', 'https://api.openai.com/v1/chat/completions')>, <Request('POST', 'https://api.openai.com/v1/chat/completions')>] = <pytest_httpx._httpx_mock.HTTPXMock object at 0x7fff7179c3b0>._requests_not_matched
ERROR tests/test_templates.py::test_template_basic[boo-extra_args3-None-None-Error: Cannot use -t/--template and --system together] - AssertionError: The following responses are mocked but not requested:
  - Match POST request on https://api.openai.com/v1/chat/completions
  
  If this is on purpose, refer to https://github.com/Colin-b/pytest_httpx/blob/master/README.md#allow-to-register-more-responses-than-what-will-be-requested
assert not [<pytest_httpx._request_matcher._RequestMatcher object at 0x7fff7160f0e0>]
ERROR tests/test_templates.py::test_template_basic[prompt: 'Say $hello'-extra_args4-None-None-Error: Missing variables: hello] - AssertionError: The following responses are mocked but not requested:
  - Match POST request on https://api.openai.com/v1/chat/completions
  
  If this is on purpose, refer to https://github.com/Colin-b/pytest_httpx/blob/master/README.md#allow-to-register-more-responses-than-what-will-be-requested
assert not [<pytest_httpx._request_matcher._RequestMatcher object at 0x7fff71496a50>]
====== 3 failed, 181 passed, 933 warnings, 3 errors in 186.32s (0:03:06) =======

```

### Additional context

Add any other context about the problem here.

### Notify maintainers

<!--
Please @ people who are in the `meta.maintainers` list of the offending package or module.
If in doubt, check `git blame` for whoever last touched something.
-->

@aldoborrero 

### Metadata

<!-- Please insert the output of running `nix-shell -p nix-info --run "nix-info -m"` below this line -->

```
› nix-shell -p nix-info --run "nix-info -m"
 - system: `"x86_64-linux"`
 - host os: `Linux 6.11.5, NixOS, 24.11 (Vicuna), 24.11.20241025.18536bf`
 - multi-user?: `yes`
 - sandbox: `yes`
 - version: `nix-env (Nix) 2.24.9`
 - nixpkgs: `/nix/store/wb6agba4kfsxpbnb5hzlq58vkjzvbsk6-source`
 ```

---

Add a :+1: [reaction] to [issues you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[issues you find important]: https://github.com/NixOS/nixpkgs/issues?q=is%3Aissue+is%3Aopen+sort%3Areactions-%2B1-desc

--- a/pkgs/development/python-modules/llm/default.nix
+++ b/pkgs/development/python-modules/llm/default.nix
@@ -10,6 +10,7 @@
   openai,
   pip,
   pluggy,
+  puremagic,
   pydantic,
   python-ulid,
   pyyaml,
@@ -21,7 +22,7 @@
 let
   llm = buildPythonPackage rec {
     pname = "llm";
-    version = "0.16";
+    version = "0.17.1";
     pyproject = true;
 
     build-system = [ setuptools ];
@@ -32,7 +33,7 @@ let
       owner = "simonw";
       repo = "llm";
       rev = "refs/tags/${version}";
-      hash = "sha256-ew8080Lv1ObjUaGicaGrj8IXXA7rtdgcWhp41O8gfVE=";
+      hash = "sha256-6OO0SIIxChM5HRJoUM4CYGbsINmc3i+iyL/oahLHhrY=";
     };
 
     patches = [ ./001-disable-install-uninstall-commands.patch ];
@@ -43,6 +44,7 @@ let
       openai
       pip
       pluggy
+      puremagic
       pydantic
       python-ulid
       pyyaml

